<a href="https://colab.research.google.com/github/cowsilver57/sessac_test/blob/main/Chapter8_1129_%EC%88%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1유형

101번 정보(row수)가 가장 많은 상위 3차종의 price값의 각 평균값은?

In [1]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/hyundai/train.csv")
df.head()

model  year  price transmission  mileage fuelType  tax(£)   mpg  \
0   Tucson  2016  12795       Manual    36263   Diesel      30  61.7   
1      I10  2012   3995       Manual    22089   Petrol      20  61.4   
2      I30  2019  24000       Manual      555   Petrol     145  34.0   
3   Tucson  2017  12995       Manual    32147   Diesel      30  61.7   
4   Tucson  2018  14991    Semi-Auto    32217   Diesel     150  57.6   

   engineSize  
0         1.7  
1         1.2  
2         2.0  
3         1.7  
4         1.7

In [6]:
answer =df.loc[df.model.isin(df.model.value_counts().index[:3])].groupby('model').mean()['price'].to_frame()
display(answer)

<ipython-input-6-137fde488720>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  answer =df.loc[df.model.isin(df.model.value_counts().index[:3])].groupby('model').mean()['price'].to_frame()


price
model                
 I10      7646.137891
 I30     11651.821759
 Tucson  15805.441373

102번 Outcome 값에 따른 각 그룹의 각 컬럼의 평균 차이를 구하여라

In [7]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/diabetes/train.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            3      102             74              0        0  29.5   
1            2      144             58             33      135  31.6   
2            5      136             82              0        0   0.0   
3           13      145             82             19      110  22.2   
4            1      117             60             23      106  33.8   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.121   32        0  
1                     0.422   25        1  
2                     0.640   69        0  
3                     0.245   57        0  
4                     0.466   27        0

In [8]:
result = df.groupby('Outcome').mean().diff().iloc[1,:]
print(result)

Pregnancies                  1.574159
Glucose                     29.943995
BloodPressure                3.202079
SkinThickness                2.952033
Insulin                     33.398645
BMI                          4.541437
DiabetesPedigreeFunction     0.143646
Age                          5.866939
Name: 1, dtype: float64


103번 매년 5월달의 open가격의 평균값을 데이터 프레임으로 표현하라

In [9]:
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nflx/NFLX.csv")
df.head()

Date      Open      High       Low     Close  Adj Close     Volume
0  2002-05-23  1.156429  1.242857  1.145714  1.196429   1.196429  104790000
1  2002-05-24  1.214286  1.225000  1.197143  1.210000   1.210000   11104800
2  2002-05-28  1.213571  1.232143  1.157143  1.157143   1.157143    6609400
3  2002-05-29  1.164286  1.164286  1.085714  1.103571   1.103571    6757800
4  2002-05-30  1.107857  1.107857  1.071429  1.071429   1.071429   10154200

In [13]:
df['Date'] = pd.to_datetime(df['Date'])
target = df.groupby(df['Date'].dt.strftime('%Y-%m')).mean()
result = target.loc[target.index.str.contains('-05')].Open
print(result)

Date
2002-05      1.155833
2003-05      1.641497
2004-05      4.261143
2005-05      1.951905
2006-05      4.163571
2007-05      3.159351
2008-05      4.435034
2009-05      5.790571
2010-05     14.417071
2011-05     34.650272
2012-05     10.523247
2013-05     31.936429
2014-05     51.121292
2015-05     85.057429
2016-05     92.705715
2017-05    158.255455
2018-05    329.779541
2019-05    359.664548
2020-05    433.880499
2021-05    496.923996
Name: Open, dtype: float64


<ipython-input-13-c15bae53ca85>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  target = df.groupby(df['Date'].dt.strftime('%Y-%m')).mean()


In [15]:
pd.set_option('display.max_columns',50)
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/nba/nba.csv",encoding='latin',sep=';')
df.head()

Rk             Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%   3P  \
0   1   Precious Achiuwa   C   22  TOR  73  28  23.6  3.6   8.3  0.439  0.8   
1   2       Steven Adams   C   28  MEM  76  75  26.3  2.8   5.1  0.547  0.0   
2   3        Bam Adebayo   C   24  MIA  56  56  32.6  7.3  13.0  0.557  0.0   
3   4       Santi Aldama  PF   21  MEM  32   0  11.3  1.7   4.1  0.402  0.2   
4   5  LaMarcus Aldridge   C   36  BRK  47  12  22.3  5.4   9.7  0.550  0.3   

   3PA    3P%   2P   2PA    2P%   eFG%   FT  FTA    FT%  ORB  DRB   TRB  AST  \
0  2.1  0.359  2.9   6.1  0.468  0.486  1.1  1.8  0.595  2.0  4.5   6.5  1.1   
1  0.0  0.000  2.8   5.0  0.548  0.547  1.4  2.6  0.543  4.6  5.4  10.0  3.4   
2  0.1  0.000  7.3  12.9  0.562  0.557  4.6  6.1  0.753  2.4  7.6  10.1  3.4   
3  1.5  0.125  1.5   2.6  0.560  0.424  0.6  1.0  0.625  1.0  1.7   2.7  0.7   
4  1.0  0.304  5.1   8.8  0.578  0.566  1.9  2.2  0.873  1.6  3.9   5.5  0.9   

   STL  BLK  TOV   PF   PTS  
0  0.5  0.6  1.2  2.1   9.1  
1  0.9  0.8  1.5  2.0   6.9  
2  1.4  0.8  2.6  3.1  19.1  
3  0.2  0.3  0.5  1.1   4.1  
4  0.3  1.0  0.9  1.7  12.9

104번 Tm 컬럼은 각 팀의 이름을 의미한다. TOR팀의 평균나이를 소수 4째 자리까지 구하여라

In [16]:
result = round(df[df.Tm=='TOR'].Age.mean(),4)
print(result)

24.8696


105번 Pos 컬럼은 포지션을 의미한다. 전체 선수 중 최소나이대의 선수들을 필터하고 그들 중 가장 많은 포지션은 무엇인지 확인하라

In [18]:
result = df[df.Age==df.Age.min()].Pos.value_counts().index[0]
print(result)

SG


106번 선수들의 이름은 first_name+ 공백 + last_name으로 이루어져 있다. 가장 많은 first_name은 무엇이며 몇 회 발생하는지 확인하라

In [26]:
result = df['Player'].str.split().str[0].str.lower().value_counts().index[0]
print(result)

justin


107번 PTS컬럼은 경기당 평균득점수 이다. 각포지션별로 경기당 평균득점수의 평균을 구하여라

In [27]:
result = df.groupby('Pos')['PTS'].mean().sort_values()
print(result)

Pos
SF        7.270253
SG-SF     7.360000
C         7.690769
PF        7.737500
SG        7.810553
PG        8.325161
PF-SF     8.500000
SF-SG     8.660000
SG-PG     9.525000
C-PF     12.850000
PG-SG    22.000000
Name: PTS, dtype: float64


108번 107번과 동일

109번 G컬럼은 참여한 경기의 숫자이다. 각 팀별로 가장 높은 경기참여 수를 가진 선수들의 경기 참여 숫자의 평균을 구하여라

In [32]:
result = df.sort_values(['Tm','G']).groupby('Tm').tail()
result2 = result.G.mean()
print(result2)

70.89032258064516


110번 Tm의 값이 MIA이며 Pos는 C또는 PF인 선수의 MP값의 평균은?

In [36]:
target = df[(df.Tm=='MIA')&(df.Pos.isin(['C','PF']))].MP.mean()
print(target)

16.7875


111번 전체 데이터중 G의 평균값의 1.5배 이상인 데이터들만 추출했을때 3P값의 평균은?

In [39]:
target = (df['G'].mean())*1.5
result = df[df.G >target]['3P'].mean()
print(result)

1.3853658536585365


112번 Age의 평균 이상인 그룹과 평균 미만인 그룹간의 G값의 평균의 차이는?

In [43]:
age_mean = df['Age'].mean()
high = df[df.Age>=age_mean]['G'].mean()
low = df[df.Age<=age_mean]['G'].mean()
print(high-low)

3.787674551781862


113번 평균 나이가 가장 젊은 팀은 어디인가

In [44]:
result = df.groupby('Tm')['Age'].mean().sort_values().index[0]
print(result)

MEM


114번 Pos 그룹별 평균 MP값을 구하여라

In [46]:
result = df.groupby('Pos')['MP'].mean()
print(result)

Pos
C        16.990000
C-PF     25.350000
PF       17.937500
PF-SF    27.300000
PG       19.547742
PG-SG    37.200000
SF       17.514557
SF-SG    20.340000
SG       18.554271
SG-PG    22.950000
SG-SF    19.620000
Name: MP, dtype: float64


#2유형

In [47]:
#github와 colab 연동
!git clone https://github.com/AnalyticsKnight/yemoonsaBigdata

Cloning into 'yemoonsaBigdata'...
remote: Enumerating objects: 637, done.
remote: Counting objects: 100% (274/274), done.
remote: Compressing objects: 100% (165/165), done.
remote: Total 637 (delta 136), reused 184 (delta 109), pack-reused 363
Receiving objects: 100% (637/637), 10.76 MiB | 6.73 MiB/s, done.
Resolving deltas: 100% (321/321), done.


In [67]:
import pandas as pd

X_test = pd.read_csv('/content/yemoonsaBigdata/datasets/Part3/304_x_test.csv')
X_train = pd.read_csv('/content/yemoonsaBigdata/datasets/Part3/304_x_train.csv')
y_train = pd.read_csv('/content/yemoonsaBigdata/datasets/Part3/304_y_train.csv')

#데이터 살펴보기
# print(X_test.head())
# print(X_train.head())
# print(y_train.head())

# print(X_test.info())
# print(X_train.info())
# print(y_train.info())

# print(X_test.isnull().sum())
# print(X_train.isnull().sum())
# print(y_train.isnull().sum())

#데이터 분류하기
COL_DEL = ['ID']
COL_NUM = ['Age','AnnualIncome','FamilyMembers','ChronicDiseases']
COL_CAT = ['Employment Type','GraduateOrNot','FrequentFlyer','EverTravelledAbroad']
COL_Y = ['TravelInsurance']

#데이터 분할하기
from sklearn.model_selection import train_test_split
X_tr, X_val, y_tr, y_val = train_test_split(X_train[COL_NUM+COL_CAT],
                                            y_train[COL_Y].values.ravel(),
                                            test_size=0.3,
                                            stratify=y_train[COL_Y].values.ravel())
# print(y_tr)

#데이터 스케일링
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_tr[COL_NUM])

X_tr[COL_NUM] = scaler.transform(X_tr[COL_NUM])
X_val[COL_NUM] = scaler.transform(X_val[COL_NUM])
X_test[COL_NUM] = scaler.transform(X_test[COL_NUM])

# print(X_tr[COL_NUM].head())

#데이터 인코딩
from sklearn.preprocessing import LabelEncoder
X = pd.concat([X_train[COL_CAT], X_test[COL_CAT]])

for col in COL_CAT:
    le = LabelEncoder()
    le.fit(X[col])
    X_tr[col] = le.transform(X_tr[col])
    X_val[col] = le.transform(X_val[col])
    X_test[col] = le.transform(X_test[col])

    # print(col)
    # print(le.classes_)
    # print('\n')

#모델 학습하기
from sklearn.ensemble import RandomForestClassifier
modelRF = RandomForestClassifier(random_state=123)
modelRF.fit(X_tr, y_tr)

from xgboost import XGBClassifier
modelXGB = XGBClassifier(random_state=123)
modelXGB.fit(X_tr, y_tr)

#예측값 생성하기
y_val_predRF = modelRF.predict_proba(X_val)[:,1]
y_val_predXGB = modelXGB.predict_proba(X_val)[:,1]

#평가 지표 구하기
from sklearn.metrics import roc_auc_score
scoreRF = roc_auc_score(y_val, y_val_predRF)
scoreXGB = roc_auc_score(y_val, y_val_predXGB)
print('Random Forest:', scoreRF)
print('XGBoost:', scoreXGB)

#함수 생성하기
def get_scores(model, X_tr, x_val, y_tr, y_val):
    y_tr_pred = model.predict_proba(X_tr)[:,1]
    y_val_pred = model.predict_proba(X_val)[:,1]
    tr_score = roc_auc_score(y_tr, y_tr_pred)
    val_score = roc_auc_score(y_val, y_val_pred)
    return f"train: {round(tr_score,4)}, valid: {round(val_score,4)}"

#성능 확인하기
print('\n')
print('Random Forest:\n', get_scores(modelRF, X_tr, X_val, y_tr, y_val))
print('XGBoost:\n', get_scores(modelXGB, X_tr, X_val, y_tr, y_val),'\n')

#과대적합으로 RF 부적절! XGBoost 사용
#하이퍼 파라미터 조정하기
modelXGB2 = XGBClassifier(n_estimators=50, max_depth=3, min_child_weight=1, random_state=123)
modelXGB2.fit(X_tr, y_tr)

modelXGB3 = XGBClassifier(n_estimators=50, max_depth=5, min_child_weight=1, random_state=123)
modelXGB3.fit(X_tr, y_tr)

modelXGB4 = XGBClassifier(n_estimators=50, max_depth=3, min_child_weight=2, random_state=123)
modelXGB4.fit(X_tr, y_tr)

modelXGB5 = XGBClassifier(n_estimators=50, max_depth=5, min_child_weight=2, random_state=123)
modelXGB5.fit(X_tr, y_tr)

print('XGBoost2:\n', get_scores(modelXGB2, X_tr, X_val, y_tr, y_val))
print('XGBoost3:\n', get_scores(modelXGB3, X_tr, X_val, y_tr, y_val))
print('XGBoost4:\n', get_scores(modelXGB4, X_tr, X_val, y_tr, y_val))
print('XGBoost5:\n', get_scores(modelXGB5, X_tr, X_val, y_tr, y_val))

Random Forest: 0.8230669265756985
XGBoost: 0.8435347628330084


Random Forest:
 train: 0.9904, valid: 0.8231
XGBoost:
 train: 0.978, valid: 0.8435 

XGBoost2:
 train: 0.8943, valid: 0.834
XGBoost3:
 train: 0.9483, valid: 0.8405
XGBoost4:
 train: 0.8837, valid: 0.8377
XGBoost5:
 train: 0.9302, valid: 0.8419


In [72]:
#최종 결과값 생성하기
pred = modelXGB4.predict_proba(X_test[COL_NUM+COL_CAT])[:,1]
result = pd.DataFrame({'ID':X_test.ID, 'pred':pred})
print(result.head())
#result.to_csv('007000955.csv', index=False)

     ID      pred
0  1569  0.168895
1  1344  0.943382
2  1429  0.199335
3   896  0.262750
4   101  0.963100


#3유형

##1. 단일표본 T-검정

In [79]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

a = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/trees.csv')
X = a['Height'].mean()
print(round(X,2))

t_score, p_value = stats.ttest_1samp(a['Height'],75)
print(round(t_score,2))

print(round(p_value,4))
if p_value>0.05:
    print('채택')
else:
    print('기각')

76.0
0.87
0.3892
채택


##2. 독립표본 T-검정

In [84]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

a1 = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/toothgrowth.csv')

stats.levene(a1.loc[a1.supp=='VC','len'], a1.loc[a1.supp=='OJ','len'])

VC_a1 = a1.loc[a1.supp=='VC']
OJ_a1 = a1.loc[a1.supp=='OJ']

VC_X = VC_a1['len'].mean()
OJ_X = OJ_a1['len'].mean()

print(round(VC_X,2), round(OJ_X,2))

t_score, p_value = stats.ttest_ind(VC_a1['len'], OJ_a1['len'], equal_var=True)
print(round(t_score,2))

print(round(p_value,4))
if p_value>0.05:
    print('채택')
else:
    print('기각')

16.96 20.66
-1.92
0.0604
채택


##3. 쌍체표본 T-검정

In [91]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

b = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/insectsprays.csv')

diff = b['after_spr']-b['before_spr']
m_d = diff.mean()
print(round(m_d,2))

t_score, p_value = stats.ttest_rel(b['before_spr'], b['after_spr'], alternative='greater')
print(round(t_score,2))

print(round(p_value,4))
if p_value>0.05:
    print('채택')
else:
    print('기각')

-12.0
14.89
0.0
기각


##4. 일원분산분석

In [99]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

b1 = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/iris.csv')

result = b1.loc[b1['species']=='versicolor','sepal_width'].mean()
print(round(result,2))

X1 = b1.loc[b1['species']=='setosa','sepal_width']
X2 = b1.loc[b1['species']=='versicolor','sepal_width']
X3 = b1.loc[b1['species']=='virginica', 'sepal_width']

test_result = stats.f_oneway(X1, X2, X3)
print(round(test_result.statistic,2))

p_value = test_result.pvalue
print(round(p_value,4))

if p_value>0.05:
    print('채택')
else:
    print('기각')

2.77
49.16
0.0
기각


##5. 카이제곱 적합도 검정

In [110]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

c = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/cellphone.csv')

count_df = c.groupby('제조사').count()
print(count_df)

result = count_df.loc['삼성','id'] - count_df.loc['애플','id']
print(result)

total_sum = len(c)
expected = [int(total_sum*0.1),
            int(total_sum*0.6),
            int(total_sum*0.3)]
print(expected)

observed = count_df['id'].values
test_result = stats.chisquare(observed, f_exp = expected)
result = round(test_result.statistic,2)
print(result)

p_value = test_result.pvalue
print(round(p_value,4))

if p_value>0.05:
    print('채택')
else:
    print('기각')

      id
제조사     
기타     5
삼성   117
애플    78
39
[20, 120, 60]
16.73
0.0002
기각


##6. 카이제곱 독립성 검정

In [117]:
import pandas as pd
import scipy.stats as stats
from math import sqrt

d = pd.read_csv('/content/yemoonsaBigdata/datasets/supplement/student.csv')
table = pd.crosstab(d['단과대학'], d['수강 과목 수'])
print(table)

result = table[table['6개 이상']==table['6개 이상'].max()].index[0]
print(result)

X1 = table.loc['경영대학',:]
X2 = table.loc['자연대학',:]
X3 = table.loc['공과대학',:]

test_result = stats.chi2_contingency([X1,X2,X3])
print(round(test_result.statistic,2))

p_value = test_result.pvalue
print(round(p_value,4))

if p_value>0.05:
    print('채택')
else:
    print('기각')

수강 과목 수  4개 이하  5개  6개 이상
단과대학                     
경영대학        12  78     30
공과대학        30  42     28
자연대학        30  42      8
경영대학
30.13
0.0
기각
